### PR TITLE
application: serial_lte_modem: Enhance #XSLEEP handling

### PR DESCRIPTION
**Pull request by junqingzou**
Rename CONFIG_SLM_INTERFACE_PIN to CONFIG_SLM_WAKEUP_PIN;
Add CONFIG_SLM_INDICATE_PIN and CONFIG_SLM_INDICATE_TIME_MS;
Deprecate original SHUTDOWN_MODE_IDLE (0);
Rename SHUTDOWN_MODE_UART(2) to SHUTDOWN_MODE_IDLE(2);
When started by unexpected reset, send an indication then start;
When downlink data arrives at Idle mode, send an indication;
Speed up wakeup GPIO callback handling using no KWork;
Send buffered downlink data after exiting Idle;
Disable UART_2 by default;

The main() functions is completely reworked.

JIRA ticket NCSIDB-704
JIRA ticket NCSDK-14085

Signed-off-by: Jun Qing Zou <jun.qing.zou@nordicsemi.no>

### CODE REVIEW

File: applications/serial_lte_modem/Kconfig [hunk 1]
```diff
@@ -81,8 +81,16 @@ config SLM_START_SLEEP
...
 	  Put nRF9160 into deep sleep mode after startup.
 
-config SLM_INTERFACE_PIN
-	int "Interface GPIO to wake up or exit idle mode"
+config SLM_WAKEUP_PIN
+	int "Interface GPIO to exit from sleep or idle mode"
+
+config SLM_INDICATE_PIN
+	default -1
+	int "Interface GPIO to indicate data available or unexpected reset"
```
Comment (tejlmand) [inline applications/serial_lte_modem/Kconfig:89]: it's a good custom to have the type (and title) of the setting at top.

Also, the title is a bit long, some of the title more belongs in the help text of the setting.

Proposal:
```suggestion
	int "SLM indication pin"
	default -1
	help
	  Interface GPIO pin used by Serial LTE Modem to indicate when data is available or an unexpected reset has occured
```

File: applications/serial_lte_modem/Kconfig [hunk 2]
```diff
@@ -81,8 +81,16 @@ config SLM_START_SLEEP
...
+
+config SLM_INDICATE_TIME
+	default 100
+	int "Indicate GPIO active time in milliseconds"
```
Comment (tejlmand) [inline applications/serial_lte_modem/Kconfig:93]: it's a good custom to have the type (and title) of the setting at top.

Also, the title is a bit long, some of the title more belongs in the help text of the setting.

Proposal:
```suggestion
	int "SLM indication period"
	default 100
	help
	  GPIO active indication time in milliseconds. This setting specify the period length for the pin to be active.
```

File: applications/serial_lte_modem/Kconfig [hunk 3]
```diff
@@ -81,8 +81,16 @@ config SLM_START_SLEEP
 	help
 	  Put nRF9160 into deep sleep mode after startup.
 
-config SLM_INTERFACE_PIN
-	int "Interface GPIO to wake up or exit idle mode"
+config SLM_WAKEUP_PIN
+	int "Interface GPIO to exit from sleep or idle mode"
```
Comment (tejlmand) [inline applications/serial_lte_modem/Kconfig:85]: Why is there no default here ?
https://docs.zephyrproject.org/latest/guides/build/kconfig/tips.html#redundant-defaults
> Defaults should always be given for int and hex symbols, however, as they implicitly default to the empty string. 

Also, the prompt text looks more like help text than a title.
Proposal:
```suggestion
	int "SLM wakeup pin"
	help
	  Interface GPIO pin used by Serial LTE Modem to exit from sleep or idle mode
```

**Review by tejlmand — CHANGES_REQUESTED** [review]
some minor requests regarding the Kconfig.

As this is sample specific setting, the change from `SLM_INTERFACE_PIN` to `SLM_WAKEUP_PIN` is acceptable, but in general we should remember that users might have adopted this app to own board, and having to adjust to n-renamed symbols can be quite annoying.
So always consider if the rename is important and / or if old symbol should be deprecated (and kept for a release or two) before completely removed.

**Comment by junqingzou** [comment]
> 

@tejlmand Updated Kconfig as advised. Thanks!

File: applications/serial_lte_modem/Kconfig (hunk 3 above)
Comment (tejlmand) [reply applications/serial_lte_modem/Kconfig]: you missed this part of my comment:
> Why is there no default here ?
> https://docs.zephyrproject.org/latest/guides/build/kconfig/tips.html#redundant-defaults
>
> > Defaults should always be given for int and hex symbols, however, as they implicitly default to the empty string.

the `int` setting doesn't have a default, which means it becomes an empty string which has the potential to cause spurious errors in code:
https://github.com/nrfconnect/sdk-nrf/blob/a63ce76f1205db154ca54eb703fc2e690ca2dde6/applications/serial_lte_modem/Kconfig#L84-L87

Please give an explanation and comment in code if there is a special reason this should not have a default.

Comment (junqingzou) [reply applications/serial_lte_modem/Kconfig]: Sorry for overlooking. We have configured the value in Zephyr project file.
But now I added -1 as default and also Build check that this value must be equal or greater than 0.
Thanks for pointing it out

**Review by rlubos — APPROVED** [review]

**Review by pirun — APPROVED** [review]
Verified indicate pin behavior with logic analyzer.
Verified power consumption with ppk2.

**Review by FrancescoSer — APPROVED** [review]

**Comment by rlubos** [comment]
The commit message body has also a line length limit (72 char/line).

**Pull request merged**